Raise interpretable error if wrong tokenizer was used when preparing the dataset

rasbt opened the pull request:
This raises an error if a user used a wrong checkpoint to prepare the dataset. Tested this approach on a bunch of models, and it seemed to catch all issues.

Sure there may be some tokenizers that produce the same number of tokens, and that won't be caught, but it catches most of the common user errors for now. 

Fixes #428
Fixes #454
Fixes #366
Fixes #478

carmocca commented on finetune/adapter.py:
```diff
@@ -222,13 +224,22 @@ def train(
     losses = torch.zeros(eval_iters)
     for k in range(eval_iters):
         input_ids, targets = get_batch(fabric, val_data, longest_seq_length)
+        if (input_ids > padded_vocab_size).any():
```
Two things here.

- You can do `model.config.padded_vocab_size` and avoid passing it across all functions
- This creates a branch on the value of input_ids, which will be expensive under compilation. This could slow down validation under that setting. Perhaps it would be better doing this only during the sanity check

This comment applies to the other scripts too

Andrei-Aksionov replied:
@carmocca 
A question: something like:
```python
if (input_ids.detach() > padded_vocab_size).any():
  ...
```
will not work?

> When we perform operations on a tensor, PyTorch builds a computation graph that tracks the operations performed on the tensor. **The detach() function is used to remove a tensor from the computation graph, making it a standalone tensor that is no longer linked to the graph**.

Yes, it will create a new tensor but `input_ids` should be quite small anyway so, in theory, the impact will be negligible.

rasbt replied:
> You can do model.config.padded_vocab_size and avoid passing it across all functions

Ah yes, this makes things easier! Thanks

> This creates a branch on the value of input_ids, which will be expensive under compilation. This could slow down validation under that setting. Perhaps it would be better doing this only during the sanity check

@Andrei-Aksionov 

I think the computation graph is not the issue, it should already be detached because the `validate` function runs in a `torch.no_grad()` context. 

When I understood the @carmocca correctly, it's just that for loops get expanded, no matter whether the computation is part of the networks computation graph, when a model is compiled? I honestly don't know the details of what gets compiled or not.

carmocca replied:
It means that you cannot fuse operations because you need to stop and look at the precise values in the tensor to know which branch to take. This is not the end of the world, but since it only needs to be checked once, I suggest that it's done during the sanity checking call

Andrei-Aksionov replied:
>  it should already be detached because the validate function runs in a torch.no_grad() context.

It looks like there is a difference between `detach` and `torch.no_grad`:
https://saturncloud.io/blog/difference-between-detach-and-with-torchnograd-in-pytorch/

> It means that you cannot fuse operations because you need to stop and look at the precise values in the tensor to know which branch to take. 

Oh, that branching! :)
I initially thought about Computational Graph branching, while you've meant [branch prediction](https://en.wikipedia.org/wiki/Branch_predictor).

> since it only needs to be checked once, I suggest that it's done during the sanity checking call

Agree with that. And it is a more appropriate place for it.

carmocca reviewed (commented):
Shouldn't pretraining also check this?

rasbt commented:
Ok, refactored that to 

1) only run it in the beginning as suggested by @carmocca 

2) put it into `utils.py`. Note that it depends on `get_batch`, so I moved that there too. The `get_batch` function is the same for adapter, adapter_v2, lora, and full finetuning. So in the spirit of @awaelchli 's suggestion, it made sense to move it there, too.

Only applied the changes to LoRA for now to get some initial feedback. What do you think?

rasbt commented on finetune/lora.py:
```diff
@@ -253,13 +257,22 @@ def train(
+    model: GPT, 
+    val_data: List[Dict], 
+    tokenizer: Tokenizer, 
+    longest_seq_length: int,
```
good point, done!

awaelchli commented on finetune/lora.py:
```diff
@@ -253,13 +257,22 @@ def train(
     losses = torch.zeros(eval_iters)
     for k in range(eval_iters):
         input_ids, targets = get_batch(fabric, val_data, longest_seq_length)
+        if (input_ids > padded_vocab_size).any():
```
I think there was a misunderstanding. My proposal was to only move the error checking there. The reason I wrote down the long explanation is to exactly avoid the confusion between "code is duplicated, therefore disassemble the scripts into a common trainer utility" vs. making the scrips self contained as much as possible. 

If it was between this and the previous version, I would strongly prefer the previous.

rasbt replied:
I see. The problem is that the error checking requires loading batches, and then we can't have the error checking function in utils because each it would then have to import `get_batch` from the individual scripts. So, maybe we should duplicate the error checking code in the individual scripts similar to `get_batch`?

awaelchli commented on lit_gpt/utils.py:
```diff
@@ -24,6 +25,61 @@ def find_multiple(n: int, k: int) -> int:
+
+
+@torch.no_grad()
+def check_compatible_tokenization(
```
There is no need to write a decoration like this. `get_batch` and the check are independent things. 

How about this:

```py
def _check_compatible_tokenization(ids, vocab_size):
 
    if (ids > ).any():
        raise ValueError(
            "Some input IDs exceed the vocabulary size of this model. "
            "Please make sure you used this model's "
            "tokenizer to prepare the dataset."
    )
```

Simply calling this in a one-liner inside get_batch would be sufficient.

rasbt replied:
You are 100% right. That was dumb, have been staring at screens too long this week 🤦‍♂️

rasbt replied:
ah wait, the problem is that we don't want to have this in the `validate` function, because this function gets also called later, and that would then be not ideal performance-wise. I think the idea was decoupling it from `validate`

carmocca commented:
Closing. This fix needs to be done the proper way, meaning resolving #471
